crates/marine-test-macro-impl/src/marine_test/modules_linker.rs

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2021 Fluence Labs Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use crate::marine_test::config_utils::Module;
+use crate::{TResult, TestGeneratorError};
+
+use marine_it_parser::it_interface::IRecordTypes;
+use marine_it_parser::it_interface::it::{IType, IRecordType};
+
+use itertools::zip;
+use std::cmp::Ordering;
+use std::collections::HashMap;
+use std::hash::Hasher;
+use std::rc::Rc;
+use static_assertions::const_assert;
+
+pub(super) fn link_modules<'modules>(
+    modules: &'modules [Module<'_>],
+) -> TResult<LinkedModules<'modules>> {
+    let mut all_record_types = HashMap::<IRecordTypeClosed<'_>, &str>::new();
+    let mut linked_modules = HashMap::<&str, LinkedModule<'_>>::new();
+
+    for module in modules {
+        let mut linking_module = LinkedModule::default();
+        for (_, record_type) in &module.interface.record_types {
+            let record_type_ex =
+                IRecordTypeClosed::new(record_type.clone(), &module.interface.record_types);
+
+            let entry = match all_record_types.get(&record_type_ex) {
+                Some(owner_module) => RecordEntry::Use(UseDescription {
+                    from: owner_module,
+                    name: &record_type.name,
+                }),
+                None => {
+                    all_record_types.insert(record_type_ex.clone(), module.name);
+                    RecordEntry::Declare(record_type_ex)
+                }
+            };
+
+            linking_module.records.push(entry);
+        }
+
+        if let Some(_) = linked_modules.insert(module.name, linking_module) {
+            return Err(TestGeneratorError::DuplicateModuleName(
+                module.name.to_string(),
+            ));
+        }
+    }
+
+    Ok(linked_modules)
+}
+
+struct ITypeClosed<'r> {
+    ty: &'r IType,
+    records: &'r IRecordTypes,
+}
+
+impl<'r> ITypeClosed<'r> {
+    fn new(ty: &'r IType, records: &'r IRecordTypes) -> Self {
+        Self { ty, records }
+    }
+}
+
+impl PartialEq for ITypeClosed<'_> {
+    fn eq(&self, other: &Self) -> bool {
+        use IType::*;
+        // check if new variants require special handling in the match below
+        const_assert!(IType::VARIANT_COUNT == 17);
```

Comment:
please introduce a new const local variable that will represent this magic number